src/tools/continuousintegration.rst

```diff
@@ -89,6 +90,132 @@ To do that the option must be enabled:
+SKA Pipelines are viewable and executable at https://gitlab.com/ska-telescope
+
+General Notes
+_____________
+- Every commit could potentially trigger a pipeline build. There may be different rules applied to determine which stages are executed in the pipeline based on factors like the branch name.
+
+    - E.g Every commit in a feature branch may trigger the “Lint” stage, but not a slow test suite.
+- When doing a release with a git tag, the full pipeline will be run.
+- Every pipeline job is associated with its git commit (including tag commits).
+- Try and have the stages complete as fast as possible.
+
+    - In some cases it may be possible to parallelize jobs. For example, unit tests and static analysis could be run in parallel.
+- If a job fails, the pipeline will be aborted and the cause of the failure needs to be addressed.
+- All projects must include all the stages listed below.
+- Project dependencies must be stored in, and made available from the SKA software repository.
+- All tests must pass on the “master” branch and should be kept stable.
+- Every stage should have some kind of JUnit output to indicate success or failure. In the case of failure it should display the error messages.
+
+Stages
+______
+Build
+"""""
+The build stage packages/compiles the software project into distributable units of software.
+The project will be checked out at the git commit hash. This specific version of the code must then be built. Failing the build stage will stop the further steps from being executed. Where possible Semantic Versioning should be used.
+To create a release a git tag should be used. `See Release Procedure <http://developer.skatelescope.org/en/latest/development/python_package_release_procedure.html>`_.
+
+Input
+  Git commit hash (provided by Gitlab)
+
+Output
+  A distributable unit of software. E.g .deb .whl .jar or docker image.
+  These must be stored as part of the artifacts and will then be available to subsequent jobs.
+  One could also store metadata together with the artefact, such as a hash of the binary artefact. This should be provided by our artefact registry.
+
+
+Linting
+"""""""
+The static analysis stage does static code analysis on the source code such as Linting.
+
+Input
+  None
+
+Output
+  Quality analysis results in JUnit format.
+
+Test
+""""
+The test stage must install/make use of the packages created during the build stage and execute tests on the installed software. Tests should be grouped into Fast / Medium / Slow / Very Slow categories.
+
+Input
+  The output from the Build stage. E.g .deb or .whl or docker image.
+  Input could also consist of test data or environment.
+
+Output
+  - The results of the tests in JUnit format. These need to be added to the artifacts.
+    `See Gitlab Test Reports <https://docs.gitlab.com/ee/ci/junit_test_reports.html>`_.
+  - Coverage metrics in JUnit format.
+
+Test types
```

Comment:
@giobra this is still a work in progress and this is not intended to be a comprehensive description of possible test types. Can you have a look at this? Do you think it makes sense to have it here?

Reply:
Sorry for jumping in so late. I was not aware of this.

Sounds good. We might add some detail such as "very fast ==> each test takes no more than a handful of ms". 
What is a "component"? a dockerized service/tool?

Why components are not mocked in integration tests? In fact, if you want to really test only the interace between 2 components, then you better mock everything other than the interface code.

Deployment tests: are like smoke tests, aka semanitc monitoring tests? if so some more details would be needed to let programmers understand what this stage is about. 



We might discuss these things in more detail within the tetsing CoP (f2f meeting at Pune?) also in the context of the SKA testing strategy. The testing strategy infact should specify all the relevant test levels/stages and provide definitions and explanations.

Reply:
thanks @giobra. Good, I guess we will provide the scaffolding for now and the detailed definitions will be filled in with the help of the CoP. 
@SKAJohanVenter can you adress the comments abour "very fast" and maybe add some `..todo` directives where you see further definition will be needed. I still think I'd like to have the page published.

Reply:
Thanks @giobra and @flyingfrog81 
I updated the docs accordingly.